Upgrade Monaco

limitedmage opened the pull request:
Fixes #19935

Tested scenarios:

- MonacoQueryEditor
  - [x] Autocomplete
  - [x] Syntax highlighting
  - [x] Keyboard shortcut (Ctrl+space)
  - [x] Hover tooltips
  - [x] Onboarding tour
- User settings JSON editor
  - [x] Autocomplete
  - [x] Syntax highlighting
  - [x] Keyboard shortcuts (Ctrl+space, Cmd+S)
  - [x] Hover tooltips
  - [x] Clicking buttons over the editor works
- Admin site configuration JSON editor
  - [x] Autocomplete
  - [x] Syntax highlighting
  - [x] Keyboard shortcuts (Ctrl+space, Cmd+S)
  - [x] Hover tooltips
  - [x] Clicking buttons over the editor works
- Admin global settings JSON editor
  - [x] Autocomplete
  - [x] Syntax highlighting
  - [x] Keyboard shortcuts (Ctrl+space, Cmd+S)
  - [x] Hover tooltips
  - [x] Clicking buttons over the editor works
  
Also changed:
- Fixed a bug where the cursor in the settings would have a focus ring. This was a small square with the old version of Monaco but is now a box the size of the cursor.
  
Not testing:
- Search console page (still experimental and disabled via feature flag)

Bugs in query input to fix before checking in:
- [x] Blank completion box appears if typing too fast (blocked on Monaco bug: https://github.com/microsoft/monaco-editor/issues/2437)
- [x] Wrong icons are shown for some categories
- [x] Logspam when rendering custom icons
- [x] Input visible box gets smaller if too much is types
- [x] Fix "Tab" tooltip and readmore icon

rvantonder commented on client/web/src/search/input/MonacoQueryInput.tsx:
```diff
@@ -131,24 +135,6 @@ const hasKeybindingService = (
     typeof (editor._standaloneKeybindingService as MonacoEditorWithKeybindingsService['_standaloneKeybindingService'])
         .addDynamicKeybinding === 'function'
 
-/**
- * HACK: this interface and the below type guard are used to add a custom command
- * to the editor. There is no public API to add a command with a specified ID and handler,
- * hence we need to use the private _commandService API.
- *
- * See upstream issue:
- * - https://github.com/Microsoft/monaco-editor/issues/900#issue-327455729
- * */
-interface MonacoEditorWithCommandService extends Monaco.editor.IStandaloneCodeEditor {
-    _commandService: {
-        addCommand: (command: { id: string; handler: () => void }) => void
-    }
-}
-
-const hasCommandService = (editor: Monaco.editor.IStandaloneCodeEditor): editor is MonacoEditorWithCommandService =>
-    hasProperty('_commandService')(editor) &&
-    typeof (editor._commandService as MonacoEditorWithCommandService['_commandService']).addCommand === 'function'
-
```
😮

lguychard replied:
SO MUCH YES

rvantonder reviewed (approved):
https://media4.giphy.com/media/cQNRp4QA8z7B6/200.gif?cid=5a38a5a2o804foywj9mvukmhez8it0cepwfkws6ltdpj67n3&rid=200.gif&ct=g

eseliger commented on client/web/src/SourcegraphWebApp.scss:
```diff
@@ -89,6 +89,7 @@ body,
 @import './search/queryBuilder/index';
 @import './searchContexts/index';
 @import './search/results/SearchResults';
+@import './settings/MonacoSettingsEditor.scss';
```
is the file ending required? Scss should be preferred over tsx 🤔

eseliger commented on client/web/src/settings/MonacoSettingsEditor.scss:
```diff
@@ -0,0 +1,9 @@
+.monaco-settings-editor {
+    .monaco-mouse-cursor-text {
+        // Hack: remove focus ring around Monaco editor cursor
```
I also noticed this when trying it out, I wonder who had this idea :D

eseliger commented on client/web/webpack.config.js:
```diff
@@ -181,6 +181,12 @@ const config = {
         include: monacoEditorPaths,
         use: ['style-loader', 'css-loader'],
       },
+      {
+        // TTF rule for monaco-editor
+        test: /\.ttf$/,
+        include: monacoEditorPaths,
+        use: ['file-loader'],
+      },
```
This rule also needs to be added to the storybook webpack config overrider here https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/storybook/src/main.ts#L138

eseliger reviewed (commented):
There is quite some logspam when typing, and the icons for the preview changed to be an art-palette 
![image](https://user-images.githubusercontent.com/19534377/114480328-a8c0b000-9c02-11eb-9fc1-cd5d4eb17ff8.png)

When typing fast, I'm sometimes able to get this weird white box:
![image](https://user-images.githubusercontent.com/19534377/114480415-d3126d80-9c02-11eb-84e0-3e721ddcd24a.png)
(But hard to reproduce, other than type a lot, fast - maybe it's been there before)
A very long query text somehow shrinks the content width and renders only a few characters https://sourcegraph.test:3443/search?q=asdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssssasdssssssssss&patternType=literal
![image](https://user-images.githubusercontent.com/19534377/114480551-081ec000-9c03-11eb-8dc5-d47b2b0db8f2.png)

Overall this looks good to me and for some of the issues I don't know if they're new, but I think we should look into these before merging, at least the content width and suggestion icons
Thanks for doing this, really really appreciated!

felixfbecker commented:
@rvantonder could you fix the Giphy link please? I was really looking forward to a GIF from you, and I feel this PR warrants one. Thank you for your consideration

felixfbecker commented:
> Fixed a bug where the cursor in the settings would have a focus ring. This was a small square with the old version of Monaco but is now a box the size of the cursor.

YES this was driving me nuts

rvantonder commented:
![asdf](https://i.giphy.com/media/cQNRp4QA8z7B6/200.gif)

rvantonder commented:
Helping to debug some artifacts, will just mention progress here. @limitedmage was saying that a blank suggestion box artifact sometimes appears:

![image (1)](https://user-images.githubusercontent.com/888624/114630410-608ba700-9c6f-11eb-9789-379155332a61.png)

I notice that this only happens when letters are being typed in very quickly. So I'm guess some type event interrupts the thing that cancels and removes the suggestion box.

rvantonder commented:
Didn't have much luck playing around with the completion providers. For the icons, the css classes have changed.

<img width="341" alt="Screen Shot 2021-04-13 at 5 48 12 PM" src="https://user-images.githubusercontent.com/888624/114638581-d6e4d500-9c80-11eb-9d51-a18eee2bdc7a.png">
<img width="569" alt="Screen Shot 2021-04-13 at 4 33 48 PM" src="https://user-images.githubusercontent.com/888624/114638585-d8ae9880-9c80-11eb-8fa1-32f6f315232d.png">

I tried setting our definitions to those, but didn't get the right result (CSS noob here). In my patch the custom icons overlap with the default ones:

<img width="229" alt="Screen Shot 2021-04-13 at 5 50 20 PM" src="https://user-images.githubusercontent.com/888624/114638824-55417700-9c81-11eb-8b25-f0bf03ce9cc0.png">

<img width="182" alt="Screen Shot 2021-04-13 at 5 50 10 PM" src="https://user-images.githubusercontent.com/888624/114638838-5b375800-9c81-11eb-924b-7f41040319db.png">

 Patch below as something to work off of.

```patch
diff --git a/client/web/src/search/input/MonacoQueryInput.scss b/client/web/src/search/input/MonacoQueryInput.scss
index d01fac8c77..bc066bf88d 100644
--- a/client/web/src/search/input/MonacoQueryInput.scss
+++ b/client/web/src/search/input/MonacoQueryInput.scss
@@ -98,3 +98,3 @@
 .theme-dark {
-    .monaco-icon-label.suggest-icon.customcolor::before {
+    .suggest-icon.codicon.codicon-symbol-property::before {
         background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGNsYXNzPSJtZGktaWNvbiAiIHdpZHRoPSIyNCIgaGVpZ2h0PSIyNCIgZmlsbD0id2hpdGUiIHZpZXdCb3g9IjAgMCAyNCA
yNCI+PHBhdGggZD0iTTE0LDEyVjE5Ljg4QzE0LjA0LDIwLjE4IDEzLjk0LDIwLjUgMTMuNzEsMjAuNzFDMTMuMzIsMjEuMSAxMi42OSwyMS4xIDE
yLjMsMjAuNzFMMTAuMjksMTguN0MxMC4wNiwxOC40NyA5Ljk2LDE4LjE2IDEwLDE3Ljg3VjEySDkuOTdMNC4yMSw0LjYyQzMuODcsNC4xOSAzLjk1LDMuNTYgNC4zOCwzLjIyQzQuNTcsMy4wOCA0Ljc4LDMgNSwzVjNIMTlWM0MxOS4yMiwzIDE5LjQzLDMuMDggMTkuNjIsMy4yMkMyMC4wNSwzLjU2IDIwLjEzLDQuMTkgMTkuNzksNC42MkwxNC4wMywxMkgxNFoiLz48L3N2Zz4=) !important;
@@ -102,3 +102,3 @@
 
-    .monaco-icon-label.suggest-icon.color::before {
+    .suggest-icon.codicon.codicon-symbol-color::before {
         background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGNsYXNzPSJtZGktaWNvbiAiIHdpZHRoPSIyNCIgaGVpZ2h0PSIyNCIgZmlsbD0id2hpdGUiIHZpZXdCb3g9IjAgMCAyNCAyNCI+PHBhdGggZD0iTTYsMkgxOEEyLDIgMCAwLDEgMjAsNFYyMEEyLDIgMCAwLDEgMTgsMjJINkEyLDIgMCAwLDEgNCwyMFY0QTIsMiAwIDAsMSA2LDJNMTIuNzUsMTMuNUMxNS41LDEzLjUgMTYuMjQsMTEuNDcgMTYuNDMsMTAuNEMxNy4zNCwxMC4xMSAxOCw5LjI2IDE4LDguMjVDMTgsNyAxNyw2IDE1Ljc1LDZDMTQuNSw2IDEzLjUsNyAxMy41LDguMjVDMTMuNSw5LjE5IDE0LjA3LDEwIDE0Ljg5LDEwLjMzQzE0LjY3LDExIDE0LDEyIDEyLDEyQzEwLjYyLDEyIDkuNjYsMTIuMzUgOSwxMi44NFY4Ljg3QzkuODcsOC41NiAxMC41LDcuNzMgMTAuNSw2Ljc1QzEwLjUsNS41IDkuNSw0LjUgOC4yNSw0LjVDNyw0LjUgNiw1LjUgNiw2Ljc1QzYsNy43MyA2LjYzLDguNTYgNy41LDguODdWMTUuMTNDNi42MywxNS40NCA2LDE2LjI3IDYsMTcuMjVDNiwxOC41IDcsMTkuNSA4LjI1LDE5LjVDOS41LDE5LjUgMTAuNSwxOC41IDEwLjUsMTcuMjVDMTAuNSwxNi4zMiA5Ljk0LDE1LjUgOS4xMywxNS4xOEM5LjQxLDE0LjUgMTAuMjMsMTMuNSAxMi43NSwxMy41TTguMjUsMTYuNUEwLjc1LDAuNzUgMCAwLDEgOSwxNy4yNUEwLjc1LDAuNzUgMCAwLDEgOC4yNSwxOEEwLjc1LDAuNzUgMCAwLDEgNy41LDE3LjI1QTAuNzUsMC43NSAwIDAsMSA4LjI1LDE2LjVNOC4yNSw2QTAuNzUsMC43NSAwIDAsMSA5LDYuNzVBMC43NSwwLjc1IDAgMCwxIDguMjUsNy41QTAuNzUsMC43NSAwIDAsMSA3LjUsNi43NUEwLjc1LDAuNzUgMCAwLDEgOC4yNSw2TTE1Ljc1LDcuNUEwLjc1LDAuNzUgMCAwLDEgMTYuNSw4LjI1QTAuNzUsMC43NSAwIDAsMSAxNS43NSw5QTAuNzUsMC43NSAwIDAsMSAxNSw4LjI1QTAuNzUsMC43NSAwIDAsMSAxNS43NSw3LjVaIi8+PC9zdmc+) !important;
@@ -108,3 +108,3 @@
 .theme-light {
-    .monaco-icon-label.suggest-icon.customcolor::before {
+    .suggest-icon.codicon.codicon-symbol-property::before {
         background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGNsYXNzPSJtZGktaWNvbiAiIHdpZHRoPSIyNCIgaGVpZ2h0PSIyNCIgZmlsbD0iIzJiMzc1MCIgdmlld0JveD0iMCAwIDI0IDI0Ij48cGF0aCBkPSJNMTQsMTJWMTkuODhDMTQuMDQsMjAuMTggMTMuOTQsMjAuNSAxMy43MSwyMC43MUMxMy4zMiwyMS4xIDEyLjY5LDIxLjEgMTIuMywyMC43MUwxMC4yOSwxOC43QzEwLjA2LDE4LjQ3IDkuOTYsMTguMTYgMTAsMTcuODdWMTJIOS45N0w0LjIxLDQuNjJDMy44Nyw0LjE5IDMuOTUsMy41NiA0LjM4LDMuMjJDNC41NywzLjA4IDQuNzgsMyA1LDNWM0gxOVYzQzE5LjIyLDMgMTkuNDMsMy4wOCAxOS42MiwzLjIyQzIwLjA1LDMuNTYgMjAuMTMsNC4xOSAxOS43OSw0LjYyTDE0LjAzLDEySDE0WiIvPjwvc3ZnPg==) !important;
@@ -112,3 +112,3 @@
 
-    .monaco-icon-label.suggest-icon.color::before {
+    .suggest-icon.codicon.codicon-symbol-color::before {
         background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGNsYXNzPSJtZGktaWNvbiAiIHdpZHRoPSIyNCIgaGVpZ2h0PSIyNCIgZmlsbD0iIzJiMzc1MCIgdmlld0JveD0iMCAwIDI0IDI0Ij48cGF0aCBkPSJNNiwySDE4QTIsMiAwIDAsMSAyMCw0VjIwQTIsMiAwIDAsMSAxOCwyMkg2QTIsMiAwIDAsMSA0LDIwVjRBMiwyIDAgMCwxIDYsMk0xMi43NSwxMy41QzE1LjUsMTMuNSAxNi4yNCwxMS40NyAxNi40MywxMC40QzE3LjM0LDEwLjExIDE4LDkuMjYgMTgsOC4yNUMxOCw3IDE3LDYgMTUuNzUsNkMxNC41LDYgMTMuNSw3IDEzLjUsOC4yNUMxMy41LDkuMTkgMTQuMDcsMTAgMTQuODksMTAuMzNDMTQuNjcsMTEgMTQsMTIgMTIsMTJDMTAuNjIsMTIgOS42NiwxMi4zNSA5LDEyLjg0VjguODdDOS44Nyw4LjU2IDEwLjUsNy43MyAxMC41LDYuNzVDMTAuNSw1LjUgOS41LDQuNSA4LjI1LDQuNUM3LDQuNSA2LDUuNSA2LDYuNzVDNiw3LjczIDYuNjMsOC41NiA3LjUsOC44N1YxNS4xM0M2LjYzLDE1LjQ0IDYsMTYuMjcgNiwxNy4yNUM2LDE4LjUgNywxOS41IDguMjUsMTkuNUM5LjUsMTkuNSAxMC41LDE4LjUgMTAuNSwxNy4yNUMxMC41LDE2LjMyIDkuOTQsMTUuNSA5LjEzLDE1LjE4QzkuNDEsMTQuNSAxMC4yMywxMy41IDEyLjc1LDEzLjVNOC4yNSwxNi41QTAuNzUsMC43NSAwIDAsMSA5LDE3LjI1QTAuNzUsMC43NSAwIDAsMSA4LjI1LDE4QTAuNzUsMC43NSAwIDAsMSA3LjUsMTcuMjVBMC43NSwwLjc1IDAgMCwxIDguMjUsMTYuNU04LjI1LDZBMC43NSwwLjc1IDAgMCwxIDksNi43NUEwLjc1LDAuNzUgMCAwLDEgOC4yNSw3LjVBMC43NSwwLjc1IDAgMCwxIDcuNSw2Ljc1QTAuNzUsMC43NSAwIDAsMSA4LjI1LDZNMTUuNzUsNy41QTAuNzUsMC43NSAwIDAsMSAxNi41LDguMjVBMC43NSwwLjc1IDAgMCwxIDE1Ljc1LDlBMC43NSwwLjc1IDAgMCwxIDE1LDguMjVBMC43NSwwLjc1IDAgMCwxIDE1Ljc1LDcuNVoiLz48L3N2Zz4=) !important;
```

limitedmage commented:
> For the icons, the css classes have changed.

Thanks @rvantonder for looking into this! I've replaced our custom icons with other icons from the VSCode icon pack. Monaco has changed from using image icons to using fonts. Also, Monaco/VSCode doesn't provide an icon for CustomColor and logs whenever one shows up (that's the logspam @esliger was seeing I think?), so I replaced our use of CustomColor with Issues instead.

Before:
![image](https://user-images.githubusercontent.com/206864/114737530-f79e4080-9cfb-11eb-9d6c-dc72386a7e86.png) ![image](https://user-images.githubusercontent.com/206864/114737547-fb31c780-9cfb-11eb-87df-2377e05f4456.png)

After:
![image](https://user-images.githubusercontent.com/206864/114737567-fff67b80-9cfb-11eb-8f29-32f2d19c14db.png) ![image](https://user-images.githubusercontent.com/206864/114737576-0258d580-9cfc-11eb-94da-70cda8d0081a.png)

FYI @rrhyne

limitedmage commented:
The "Read more" icon has also changed, causing issues with our "Tab to complete" tip. I restored the old look:
![image](https://user-images.githubusercontent.com/206864/114742040-32a27300-9d00-11eb-98a5-7f8ef83c34d3.png)

limitedmage commented:
I think I've narrowed the Monaco blank suggestion box fast typing issue to this: https://github.com/microsoft/monaco-editor/issues/2437

rvantonder commented:
Dang! Interesting. Out of curiosity, do Loic's commits work around that?

limitedmage commented:
> Dang! Interesting. Out of curiosity, do Loic's commits work around that?

They help make the issue happen less often and make typing less laggy but it still happens.

felixfbecker commented:
Lol they should've used RxJS 😄

limitedmage commented:
PR submitted to VSCode: https://github.com/microsoft/vscode/pull/121396

Question: Do we want to get this upgrade in now, or wait an unknown amount of time until this gets fixed in VSCode and released a new version of Monaco?

rvantonder commented:
In general I'd be in favor of an upstream fix (we've waited this long to upgrade Monaco). But I don't know what's blocked on this PR, and how long it would take monaco to release the update ([looks like roughly every month, but there are some rapid patch releases](https://github.com/microsoft/monaco-editor/releases))

limitedmage commented:
@sourcegraph/frontend-devs What is the biggest benefit of upgrading Monaco? Moving to Webpack 5?

eseliger commented:
Yeah, I guess that would be it. I am stuck on another blocker though, so it's not super urgent, maybe we give it a couple days and reevaluate?

limitedmage commented:
Update: Fix has been checked into VSCode, waiting to see what the process is/how long it will take to get a new Monaco release with the fix.

limitedmage commented:
Update: Monaco has been updated!! 🎉  I'll run the test scenarios again to make sure everything still works fine and then check this in.

lguychard commented:
Great stuff!

rvantonder commented:
Check up on `search/console` which continues to be in a working state, so we're good there. Thanksssssssssssssssssssssssssssss
